### PR TITLE
Support for ARM SVE2.

### PR DESCRIPTION
**Pull request by zvookin**
Heavily based on Steve Suzuki's work here: https://github.com/halide/Halide/pull/6781 . Hopefully easier to merge with less effect on existing ARM support and fewer constraints on CodeGen_LLVM.

### CODE REVIEW

**Comment by zvookin** [comment]
I can't seem to name Steve Suzuki in the reviewers list. Apologies if I'm missing something.

File: src/CodeGen_ARM.cpp [hunk 1]
```diff
@@ -104,17 +104,32 @@ class CodeGen_ARM : public CodeGen_Posix {
...
+    /** Similar to llvm_type_of, but allows forcing use of Fixed of Vscale vectors. */
+    llvm::Type *llvm_type_with_constraint(const Type &t, bool scalars_are_vectors, VectorTypeConstraint constraint);
+
+    /** Define a wrapper LLVM func that takes some arguments which Halide defines
+     * and call inner LLVM intrinsic with an additional argument which LLVM requires. */
+    llvm::Function *define_intrin_wrapper(const std::string &inner_name,
+                                          const Type &ret_type,
+                                          const std::string &mangled_name,
+                                          const std::vector<Type> &arg_types,
+                                          bool add_inactive_arg,  // for SVE
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:119]: nit: with this many bool args, it's easy to get the order wrong at a callsite. would it be silly to consider a bitfields or struct-to-simulate-call-by-named-arg?

File: src/CodeGen_ARM.cpp [hunk 2]
```diff
@@ -149,22 +174,46 @@ class CodeGen_ARM : public CodeGen_Posix {
...
+            return Shuffle::make_concat({const_true(true_lanes), const_false(false_lanes)});
+        }
+    }
 };
 
 CodeGen_ARM::CodeGen_ARM(const Target &target)
     : CodeGen_Posix(target) {
 
+    // TODO(issue needed): See if use_llvm_vector_intrinsics can replace some special
+    // handling in this file, specificall in Load and Store visitors.
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:215]: specifically

File: src/CodeGen_ARM.cpp [hunk 3]
```diff
@@ -298,22 +347,63 @@ struct ArmIntrinsic {
         SplitArg0 = 1 << 6,          // This intrinsic requires splitting the argument into the low and high halves.
         NoPrefix = 1 << 7,           // Don't prefix the intrinsic with llvm.*
         RequireFp16 = 1 << 8,        // Available only if Target has ARMFp16 feature
+        Neon64Unavailable = 1 << 9,  // Unavailalbe for 64 bit NEON
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:350]: Unavailable

File: src/CodeGen_ARM.cpp [hunk 4]
```diff
@@ -298,22 +347,63 @@ struct ArmIntrinsic {
...
+        SveUnavailable = 1 << 10,    // Unavailable for SVE
+        SveNoPredicate = 1 << 11,    // In SVE intrinsics, additional predicate argument is required as default, unless this flag is set.
+        SveInactiveArg = 1 << 12,    // This intrinsic needs the additional argument for fallback value for the lanes inactivated by predicate.
+        SveRequired = 1 << 14,       // This intrinsic requires SVE.
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:354]: nit: are we skipping 13?

File: src/CodeGen_ARM.cpp [hunk 5]
```diff
@@ -705,32 +823,101 @@ const std::map<string, string> float16_transcendental_remapping = {
...
+        // We don't use this, so just pad it with 0.
+        inner_llvm_arg_types.push_back(llvm_ret_type);
+
+        Value *zero = Constant::getNullValue(llvm_ret_type);
+        inner_args.push_back(zero);
+    }
+    if (add_predicate) {
+        llvm::Type *pred_type = to_llvm_type(Int(1, inner_lanes));
+        inner_llvm_arg_types.push_back(pred_type);
+        // For now, we don't use predicate in overloaded intrinsic
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:879]: Do we need an issue for this?

File: src/CodeGen_ARM.cpp [hunk 6]
```diff
@@ -861,8 +1114,30 @@ void CodeGen_ARM::compile_func(const LoweredFunc &f,
     CodeGen_Posix::compile_func(func, simple_name, extern_name);
 }
 
+void CodeGen_ARM::begin_func(LinkageType linkage, const std::string &simple_name,
+                             const std::string &extern_name, const std::vector<LoweredArgument> &args) {
+    CodeGen_Posix::begin_func(linkage, simple_name, extern_name, args);
+
+    // TODO(issue neeed): There is likely a better way to ensure this is only
+    // genrated for the outermost function that is being compiled. Avoiding the
```
Comment (steven-johnson) [inline src/CodeGen_ARM.cpp:1122]: generated

**Comment by steven-johnson** [comment]
looks like legit failure in correctness_vector_reductions on arm64

File: src/CodeGen_ARM.cpp [hunk 7]
```diff
@@ -104,17 +104,32 @@ class CodeGen_ARM : public CodeGen_Posix {
     CodeGen_ARM(const Target &);
 
 protected:
+    using codegen_func_t = std::function<Value *(int lanes, const std::vector<Value *> &)>;
     using CodeGen_Posix::visit;
 
-    /** Assuming 'inner' is a function that takes two vector arguments, define a wrapper that
-     * takes one vector argument and splits it into two to call inner. */
-    llvm::Function *define_concat_args_wrapper(llvm::Function *inner, const string &name);
+    /** Similar to llvm_type_of, but allows forcing use of Fixed of Vscale vectors. */
```
[the remark this answers is not in the document]
Comment (zvookin) [reply src/CodeGen_ARM.cpp]: Clarified comment/fixed grammatical error.

File: src/CodeGen_ARM.cpp (hunk 2 above)
Comment (zvookin) [reply src/CodeGen_ARM.cpp]: Done.

File: src/CodeGen_ARM.cpp (hunk 3 above)
Comment (zvookin) [reply src/CodeGen_ARM.cpp]: Done.

File: src/CodeGen_ARM.cpp (hunk 4 above)
Comment (zvookin) [reply src/CodeGen_ARM.cpp]: Fixed.

File: src/CodeGen_ARM.cpp (hunk 5 above)
Comment (zvookin) [reply src/CodeGen_ARM.cpp]: Changed comment.

File: src/CodeGen_ARM.cpp (hunk 6 above)
Comment (zvookin) [reply src/CodeGen_ARM.cpp]: Done.

File: src/CodeGen_ARM.cpp (hunk 1 above)
Comment (zvookin) [reply src/CodeGen_ARM.cpp]: Changed argument passing to just pass the flags directly.

**Comment by steven-johnson** [comment]
Passing all tests. How close are we to landing?

**Comment by zvookin** [comment]
Why is the IR sharing issue only affecting this SIMD op check test, not others?

**Comment by abadams** [comment]
The other simd op check tests just use Exprs and some dummy calls to functions that get subbed out with per-test-case local imageparams, so there's no shared mutable IR. This test defines Funcs that are used by multiple test cases. It's the sharing of those Funcs that caused the problem.

**Comment by abadams** [comment]
Though now that you say that, I guess we should move the deep copying code I added into the base class so that this isn't an issue elsewhere later.

**Comment by zvookin** [comment]
It should be ready to land, pending validation that it does not break anything in existing ARM support, so please run a global presubmit. We can discuss testing correctness one to one. My first attempt to test this under QEMU failed completely because FreeBSD on QEMU on Mac OS (M3) is an exercise in disk corruption.

**Comment by steven-johnson** [comment]
Looks to be clean inside Google, are we ready to land?

**Comment by abadams** [comment]
I still need to benchmark it on our workloads.

**Comment by abadams** [comment]
On average it makes no difference to our neon workloads, though neon codegen definitely changed in a bunch of places. 

I did find one minor inefficiency while eyeballing things, which uncovered an inefficiency in main. Consider this pipeline:

```
#include "Halide.h"
using namespace Halide;

int main() {
    Func f, g;
    Var x;

    f(x) = x / 17.0f;
    f.compute_root();

    g(x) = {cast<int>(ceil(f(x))), cast<int>(floor(f(x)))};

    g.vectorize(x, 4);

    g.compile_to_assembly("/dev/stdout", {}, "", Target{"arm-64-linux-no_asserts-no_runtime-no_bounds_query"});

    return 0;
}
```

On both main and the branch this generates the inner loop:
```
	ldr	q0, [x9], #16
	subs	x8, x8, #1
	frintp	v1.4s, v0.4s
	frintm	v0.4s, v0.4s
	fcvtzs	v1.4s, v1.4s
	fcvtzs	v0.4s, v0.4s
	str	q1, [x10], #16
	str	q0, [x11], #16
```

It rounds up and rounds down as a float, then converts to an int. It should instead be using a different rounding mode on the int conversions and just using fcvtms and fcvtps

If you remove the .vectorize call then on main we get:

```
	ldr	s0, [x8], #4
	subs	x19, x19, #1
	fcvtps	w9, s0
	fcvtms	w10, s0
	str	w9, [x20], #4
	str	w10, [x21], #4
```

and on the branch we get:

```
	ldr	s0, [x8], #4
	subs	x19, x19, #1
	frintp	s1, s0
	frintm	s0, s0
	fcvtzs	v1.2s, v1.2s
	fcvtzs	v0.2s, v0.2s
	st1	{ v1.s }[0], [x20], #4
	st1	{ v0.s }[0], [x21], #4
```

It looks like the branch goes down the vector path even in the scalar case, and hits the issue we already have with vector floor/ceil.

For the vector case, it looks like the llvm IR is:
```
  %22 = tail call <4 x float> @llvm.floor.v4f32(<4 x float> %18)
  %23 = fptosi <4 x float> %22 to <4 x i32>
```

but we should instead be calling the intrinsic llvm.aarch64.neon.fcvtms.v4i32.v4f32

I'll open a PR onto this branch

**Comment by abadams** [comment]
Oops, looks like I broke something. Will fix.

**Comment by abadams** [comment]
Can't seem to repro locally, and in the llvm commit log there have been reverts of aarch64 stuff. I'll just do a merge with main and see what happens.

**Review by abadams — APPROVED** [review]

**Pull request merged**